program/rust/src/time_machine_types.rs

```diff
@@ -1,45 +1,192 @@
+#![allow(unused_imports)]
+#![allow(dead_code)]
 use crate::c_oracle_header::{
     PriceAccount,
     PythAccount,
     EXTRA_PUBLISHER_SPACE,
     PC_ACCTYPE_PRICE,
+    PC_MAX_SEND_LATENCY,
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
 
-
-#[derive(Debug, Clone, Copy)]
-#[repr(C)]
-/// this wraps multiple SMA and tick trackers, and includes all the state
-/// used by the time machine
-pub struct TimeMachineWrapper {
-    //Place holder with the size of the fields I am planning to add
-    place_holder: [u8; 1864],
-}
-
+pub const THIRTY_MINUTES: i64 = 30 * 60;
+pub const NUM_BUCKETS_THIRTY_MIN: usize = 48;
 #[derive(Copy, Clone)]
 #[repr(C)]
-/// wraps everything stored in a price account
 pub struct PriceAccountWrapper {
-    //an instance of the c price_t type
+    // An instance of price account
     pub price_data:            PriceAccount,
-    //space for more publishers
+    /// Space for more publishers
     pub extra_publisher_space: [u8; EXTRA_PUBLISHER_SPACE as usize],
-    //TimeMachine
-    pub time_machine:          TimeMachineWrapper,
+    // TimeMachine
+    pub time_machine:          SmaTracker<NUM_BUCKETS_THIRTY_MIN>,
+}
+
+
+#[derive(Debug)]
+pub struct DataPoint {
+    pub last_two_timestamps: (i64, i64),
+    pub slot_gap:            u64,
+    pub price:               i64,
 }
+
 impl PriceAccountWrapper {
     pub fn initialize_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+        #[cfg(test)]
+        self.time_machine
+            .initialize(THIRTY_MINUTES, PC_MAX_SEND_LATENCY.into());
         Ok(())
     }
 
     pub fn add_price_to_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+        #[cfg(test)]
+        self.time_machine.add_datapoint( &DataPoint{
+            last_two_timestamps : (self.price_data.prev_timestamp_, self.price_data.timestamp_),
+            slot_gap : (self.price_data.last_slot_ - self.price_data.prev_slot_),
+            price: self.price_data.prev_price_ /2  + self.price_data.agg_.price_ / 2 + (self.price_data.prev_price_ % 2) * (self.price_data.agg_.price_ % 2), // Hack to avoid overflow?
+        }
+        )?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+pub struct SmaTracker<const NUM_BUCKETS: usize> {
+    /// The maximum gap in slots for an epoch's sma to be valid, we set it to PC_MAX_SEND_LATENCY
```

Comment:
get rid of "we set it to...". no reason that this needs to be specified in a code comment, because someone can just find where the struct is initialized.

Reply:
Yes